### ISSUE TITLE
"Characters remaining" in Joplin—plain text fields

### ISSUE THREAD

**Issue by desigonz** [cityofaustin/techstack#2734]
**Acceptance criteria**
For plain text fields:
- [x] Titles (all content types)
- [x] Descriptions (all content types)
- [ ] Official document content type
  - [ ] Document title
  - [ ] Authoring office of document
  - [ ] Document summary
  - [ ] Name of document
- [x] Section headings (guides)

the following should happen: 
- [x] Character limits on edit pages uses the same patterns as on the create content modal
  - [x] Displays characters remaining
  - [x] When user goes over character count, the type becomes red and displays negative number
- [x] Allows you to add more content than character limit, but with error messaging above

If the last three points become hard, let's split into two issues.

**Joplin 3.0**
- Character limits only appear for English content; in other languages, don't include (this is because translations can be longer than English; I believe we've already set the character count higher for Spanish?)
- If over the character count, can't save

-----
### Background

From #2611:

A character count pattern has been used in the content creation modal - it shows both characters left, and then switches to red type and negative number when the count goes over. 
![Screen Shot 2019-08-01 at 2 11 25 PM](https://user-images.githubusercontent.com/3790067/62320664-65da9b00-b466-11e9-8450-7de179df9ba8.png)

This pattern should be duplicated to relevant fields on page templates, for every field which has a character limit, as seen below.

![Screen Shot 2019-08-01 at 2 28 42 PM](https://user-images.githubusercontent.com/3790067/62321921-72142780-b469-11e9-838e-b1571a8b8c3e.png)

Note that for this to work correctly the required error state must be adjusted (see https://github.com/cityofaustin/techstack/issues/2672)

Character limits per field may be found in the content model google doc.

**Comment by easherma** [comment]
We didn't set the limit higher for Spanish, that isn't possible with our current setup using wagatil modeltranslations. We just raised the limits overall. 
On the subject of character limits, there is some talk in wagtail about the same issue: 
https://github.com/wagtail/wagtail/issues/2268

**Comment by desigonz** [comment]
- [ ] @desigonz to review her notes and remember where we landed on this!

**Comment by easherma** [comment]
So the way Wagtail renders plain fields is that it has the 'max_length' attribute set in the html based on the model...so that'l be pretty easy to grab via JS and do some characters remaining

**Comment by easherma** [comment]
https://github.com/wagtail/wagtail/blob/master/wagtail/utils/widgets.py <--- common pattern that Wagtail uses to pass attributes into html/JS

**Comment by easherma** [comment]
https://github.com/gasman/weps/blob/inline-script-deprecation/text/033-inline-script-deprecation.md though that might not be a good pattern to follow

**Comment by easherma** [comment]
FYI @desigonz and @tori : we should be able to implement this pretty easily as is, except the negative number aspect. It’s not possible to exceed the limit of the form element. My plan would be to implement this as is and cross that particular bridge when we get there.

**Comment by easherma** [comment]
https://docs.google.com/spreadsheets/d/1ZkH93F4tJ0_Zjx4t8k6k6TgwSlOh8u8ejWmegE6WqH0/edit#gid=1814414513 limits here

I'm not sure these limits correspond to whats actually in the DB. So for instance with TItle I _could_ actually make the negative values you want, but only because the max_length set in the database is 255 and not 58. 

This means a user could theoretically stay under the limit on the create content modal, but make the title longer later. The character limit on the edit page would be red, but it wouldn't stop them from publishing. 

However, because we want different requirements for drafts and publishing, we might end up doing our own custom validation later, so maybe this is acceptable for now and we'll eventually put in place some intermediate validation or something

**Comment by easherma** [comment]
@desigonz @toribr ready for QA:
https://joplin-pr-2734-countable.herokuapp.com/

**Comment by easherma** [comment]
Also worth noting:

We can also count limits by words or sentences, which might be better guidance for authors, especially on longer sections

The limit of 5000 for additional content seems like a red flag, if a field is that long it should probably be a rich text widget

**Comment by desigonz** [comment]
I tested against the acceptance criteria on staging. Back to you, @easherma 

- [ ] I'm able to save a page even when I've gone over the character count, when I shouldn't be able to (per acceptance criteria)
- [ ] Official document character counts aren't working; see gif
![official doc character count](https://user-images.githubusercontent.com/5839508/64299806-2826d180-cf40-11e9-84b9-5955dad6f70e.gif)

- [ ] Official document description: trippy double text 
![image](https://user-images.githubusercontent.com/5839508/64299823-32e16680-cf40-11e9-8ce2-5501c69d3345.png)

**Comment by briaguya** [comment]
![Sep-05-2019 00-39-10](https://user-images.githubusercontent.com/6384156/64314844-fcbed980-cf75-11e9-87e3-370a24d9417f.gif)

@easherma looks like @desigonz already saw this.

@desigonz re: going over character count, @easherma mentioned an issue with allowing negative characters remaining when we have a set character count (we would simply not be able to enter any more in the box past the limit)

This brings to light the question of saving drafts with too many characters, and could be addressed by https://github.com/cityofaustin/techstack/issues/2767

My recommendation is we keep the limit cosmetic for now, which gives us the ability to save too many characters in our drafts, and then we make the character limit required for publish a part of Joplin 3.

**Comment by desigonz** [comment]
@easherma Updated the QA criteria. Please fix and then send back to me for QA!

**Comment by easherma** [comment]
Yes here's more context about negative counts from the PR:
Allowing negative counts:

So this is part of the acceptance criteria
TL;DR here is, I think, the best way to approach this:
We should approach limits as a author interface UX thing and not a technical thing
For one, Django (especially using Postgres) treats the max_length as essentially cosmetic; it is not actually limiting the length of the field in the database for TextArea, and for TextInput widgets only if the underlying field is a CharField. There is not really much of a performance benefit for limiting fields.
But if you do set max_length in Django/Wagtail, the form input will literally not let you enter further characters past the limit
So I propose setting any max_length, if one is set at all, to be higher than the limit you might set with the widget. Better yet, just set the limit for that field using the widget so that it displays as intended. It sounds more complicated than it ends up being.
